File: src/hybrid_astar_planner/src/motion_model.cpp
#include "hybrid_astar_planner/motion_model.hpp"
#include <cmath>
#include <algorithm>

namespace hybrid_astar_planner
{

void MotionModel::initialize(const VehicleConfig & vehicle, const MotionConfig & motion)
{
  vehicle_ = vehicle;
  motion_ = motion;

  // Minimum turning radius: R = L / tan(delta_max)
  if (std::abs(vehicle_.max_steering_angle) > 1e-6) {
    min_turning_radius_ = vehicle_.wheelbase / std::tan(vehicle_.max_steering_angle);
  } else {
    min_turning_radius_ = 1e6;  // effectively infinite
  }

  generatePrimitives();
  generateFootprintTemplate();
}

void MotionModel::generatePrimitives()
{
  primitives_.clear();

  int n = motion_.num_steering_angles;
  double max_steer = vehicle_.max_steering_angle;

  // Generate evenly spaced steering angles
  std::vector<double> angles;
  if (n == 1) {
    angles.push_back(0.0);
  } else {
    for (int i = 0; i < n; ++i) {
      double angle = -max_steer + 2.0 * max_steer * i / (n - 1);
      angles.push_back(angle);
    }
  }

  // Forward primitives
  for (double angle : angles) {
    primitives_.push_back({Direction::FORWARD, angle});
  }

  // Reverse primitives
  if (motion_.allow_reverse) {
    for (double angle : angles) {
      primitives_.push_back({Direction::REVERSE, angle});
    }
  }
}

void MotionModel::generateFootprintTemplate()
{
  footprint_template_.clear();

  double half_w = vehicle_.width / 2.0;
  double rear = -vehicle_.rear_axle_to_back;
  double front = vehicle_.length - vehicle_.rear_axle_to_back;

  // Corner vertices
  vertex_template_[0] = {front, half_w};    // front-left
  vertex_template_[1] = {front, -half_w};   // front-right
  vertex_template_[2] = {rear, -half_w};    // rear-right
  vertex_template_[3] = {rear, half_w};     // rear-left

  // Sparse collision check points (16-20 points)
  // 4 corners
  for (const auto & v : vertex_template_) {
    footprint_template_.push_back(v);
  }

  // Edge midpoints
  footprint_template_.push_back({front, 0.0});         // front center
  footprint_template_.push_back({rear, 0.0});           // rear center
  footprint_template_.push_back({(front + rear) / 2.0, half_w});   // left center
  footprint_template_.push_back({(front + rear) / 2.0, -half_w});  // right center

  // Quarter-width points on front and rear edges
  footprint_template_.push_back({front, half_w / 2.0});
  footprint_template_.push_back({front, -half_w / 2.0});
  footprint_template_.push_back({rear, half_w / 2.0});
  footprint_template_.push_back({rear, -half_w / 2.0});

  // Side midpoints at quarter and three-quarter length
  double q1 = rear + (front - rear) * 0.25;
  double q3 = rear + (front - rear) * 0.75;
  footprint_template_.push_back({q1, half_w});
  footprint_template_.push_back({q1, -half_w});
  footprint_template_.push_back({q3, half_w});
  footprint_template_.push_back({q3, -half_w});
}

State MotionModel::applyMotion(
  const State & current,
  Direction direction,
  double steering_angle) const
{
  double step = motion_.step_size;
  double v = (direction == Direction::FORWARD) ? step : -step;

  double new_x, new_y, new_theta;

  if (std::abs(steering_angle) < 1e-6) {
    // Straight motion
    new_x = current.x + v * std::cos(current.theta);
    new_y = current.y + v * std::sin(current.theta);
    new_theta = current.theta;
  } else {
    // Arc motion using exact geometry
    double R = vehicle_.wheelbase / std::tan(steering_angle);
    double dtheta = v / R;

    new_x = current.x + R * (std::sin(current.theta + dtheta) - std::sin(current.theta));
    new_y = current.y + R * (-std::cos(current.theta + dtheta) + std::cos(current.theta));
    new_theta = normalizeAngle(current.theta + dtheta);
  }

  State result(new_x, new_y, new_theta);
  result.direction = direction;
  result.steering = steering_angle;
  return result;
}

Footprint MotionModel::getFootprint(const State & state) const
{
  double cos_t = std::cos(state.theta);
  double sin_t = std::sin(state.theta);

  Footprint fp;
  fp.points.reserve(footprint_template_.size());

  for (const auto & [lx, ly] : footprint_template_) {
    auto [wx, wy] = rotatePoint(lx, ly, cos_t, sin_t);
    fp.points.push_back({wx + state.x, wy + state.y});
  }

  return fp;
}

std::array<std::pair<double, double>, 4> MotionModel::getVertices(const State & state) const
{
  double cos_t = std::cos(state.theta);
  double sin_t = std::sin(state.theta);

  std::array<std::pair<double, double>, 4> result;
  for (int i = 0; i < 4; ++i) {
    auto [lx, ly] = vertex_template_[i];
    auto [wx, wy] = rotatePoint(lx, ly, cos_t, sin_t);
    result[i] = {wx + state.x, wy + state.y};
  }
  return result;
}

std::pair<double, double> MotionModel::rotatePoint(
  double x, double y, double cos_t, double sin_t)
{
  return {
    x * cos_t - y * sin_t,
    x * sin_t + y * cos_t
  };
}

}  // namespace hybrid_astar_planner

File: src/hybrid_astar/motion_model.py
"""
Vehicle Motion Model - Bicycle Kinematics
Based on Dolgov et al. (2010) and Meng et al. (2023)
"""

import numpy as np
from typing import List, Tuple
from src.state import State


class MotionModel:
    """
    Bicycle kinematic model for non-holonomic vehicle motion.
    Uses exact arc geometry for motion primitive generation.
    """

    def __init__(self, wheel_base: float = 2.5,
                 max_steering_angle: float = 40.0,
                 step_size: float = 0.5,
                 num_steering_angles: int = 5,
                 allow_reverse: bool = True):

        self.wheel_base = wheel_base
        self.max_steering_angle = np.radians(max_steering_angle)
        self.step_size = step_size
        self.allow_reverse = allow_reverse

        # Generate steering angles
        if num_steering_angles == 1:
            self.steering_angles = [0.0]
        else:
            self.steering_angles = np.linspace(
                -self.max_steering_angle,
                self.max_steering_angle,
                num_steering_angles
            ).tolist()

        self.directions = [1]
        if allow_reverse:
            self.directions.append(-1)

        self.min_turning_radius = wheel_base / np.tan(self.max_steering_angle)

        print(f"Motion Model: wheel_base={wheel_base}m, step={step_size}m, "
              f"steering_angles={num_steering_angles}, min_radius={self.min_turning_radius:.2f}m")

    def get_motion_primitives(self) -> List[Tuple[int, float]]:
        """Get all (direction, steering_angle) primitives."""
        primitives = []
        for direction in self.directions:
            for steering in self.steering_angles:
                primitives.append((direction, steering))
        return primitives

    def apply_motion(self, state: State, direction: int,
                     steering_angle: float) -> State:
        """
        Apply a single motion primitive using exact arc geometry.

        Args:
            state: Current state
            direction: 1 (forward) or -1 (reverse)
            steering_angle: Steering angle in radians

        Returns:
            New state after applying motion
        """
        d = self.step_size * direction

        if abs(steering_angle) < 1e-6:
            # Straight motion
            x_new = state.x + d * np.cos(state.theta)
            y_new = state.y + d * np.sin(state.theta)
            theta_new = state.theta
        else:
            # Arc motion using center-of-rotation
            R = self.wheel_base / np.tan(steering_angle)
            cx = state.x - R * np.sin(state.theta)
            cy = state.y + R * np.cos(state.theta)
            d_theta = d / R
            theta_new = state.theta + d_theta
            x_new = cx + R * np.sin(theta_new)
            y_new = cy - R * np.cos(theta_new)

        return State(x_new, y_new, theta_new,
                     direction=direction, steering=steering_angle)

    def compute_motion_cost(self, state: State, next_state: State,
                            steering_penalty: float = 1.5,
                            reversing_penalty: float = 2.0,
                            steering_change_penalty: float = 1.5,
                            direction_switch_penalty: float = 10.0) -> float:
        """
        Compute motion cost with 4 core penalties per Dolgov et al.

        Cost = step_size
             * (1 + steering_penalty * |δ|/δ_max)
             * (reversing_penalty if reverse)
             * (1 + steering_change_penalty * |Δδ|/δ_max)
             + direction_switch_penalty if gear change
        """
        cost = self.step_size

        # Proportional steering penalty: sharper turns cost more
        steer_ratio = abs(next_state.steering) / self.max_steering_angle
        cost *= (1.0 + steering_penalty * steer_ratio)

        # Reverse driving penalty
        if next_state.direction < 0:
            cost *= reversing_penalty

        # Steering angle change cost (proportional to change magnitude)
        if state.parent is not None:
            steer_diff = abs(next_state.steering - state.steering) / self.max_steering_angle
            cost *= (1.0 + steering_change_penalty * steer_diff)

        # Direction switch penalty (forward <-> reverse gear change)
        if state.direction != next_state.direction:
            cost += direction_switch_penalty

        return cost


class VehicleFootprint:
    """
    Vehicle footprint for collision checking.
    Uses sparse critical points for fast collision checks.
    """

    def __init__(self, length: float = 4.5, width: float = 2.0,
                 rear_axle_to_back: float = 1.0):

        self.length = length
        self.width = width
        self.rear_axle_to_back = rear_axle_to_back

        # 4 corners of the bounding box
        self.vertices = np.array([
            [-rear_axle_to_back, -width / 2],
            [-rear_axle_to_back, width / 2],
            [length - rear_axle_to_back, width / 2],
            [length - rear_axle_to_back, -width / 2]
        ])

        self.collision_points = self._generate_collision_points()

        print(f"Vehicle: {length}m x {width}m, collision points: "
              f"{len(self.collision_points)}")

    def _generate_collision_points(self) -> np.ndarray:
        """Generate critical points for collision checking (~16-20 points)."""
        points = []

        # 4 corners
        points.extend(self.vertices.tolist())

        # 4 midpoints of edges
        for i in range(4):
            v1 = self.vertices[i]
            v2 = self.vertices[(i + 1) % 4]
            midpoint = (v1 + v2) / 2
            points.append(midpoint.tolist())

        # Front and rear center
        front_x = self.length - self.rear_axle_to_back
        rear_x = -self.rear_axle_to_back
        points.append([front_x, 0])
        points.append([rear_x, 0])

        # Quarter-width points along length
        for x_frac in [0.25, 0.5, 0.75]:
            x_pos = rear_x + x_frac * self.length
            points.append([x_pos, -self.width / 4])
            points.append([x_pos, self.width / 4])

        return np.array(points)

    def get_footprint_fast(self, state: State) -> np.ndarray:
        """Transform collision points to world coordinates."""
        cos_theta = np.cos(state.theta)
        sin_theta = np.sin(state.theta)
        rotation = np.array([[cos_theta, -sin_theta],
                             [sin_theta, cos_theta]])

        points_world = (rotation @ self.collision_points.T).T
        points_world[:, 0] += state.x
        points_world[:, 1] += state.y

        return points_world

    def get_vertices_world(self, state: State) -> np.ndarray:
        """Get 4 corners in world coordinates for visualization."""
        cos_theta = np.cos(state.theta)
        sin_theta = np.sin(state.theta)
        rotation = np.array([[cos_theta, -sin_theta],
                             [sin_theta, cos_theta]])

        vertices_world = (rotation @ self.vertices.T).T
        vertices_world[:, 0] += state.x
        vertices_world[:, 1] += state.y

        return vertices_world
